File: sdk/typescript/src/accounts/smart-account.ts
import { CovenantSDK } from '../sdk';

export type PaymentSchedule = {
  frequency: 'daily' | 'weekly' | 'monthly';
  amount: bigint;
  token: string;
  startTime: number;
  endTime: number;
};

export type SessionKey = {
  sessionKeyAddress: string;
  sessionData: string;
  validUntil: number;
};

/**
 * Smart account extensions for the COVENANT TypeScript SDK.
 */
export class CovenantSmartAccount {
  private sdk: CovenantSDK;
  private smartAccount: any;

  constructor(sdk: CovenantSDK, smartAccount: any) {
    this.sdk = sdk;
    this.smartAccount = smartAccount;
  }

  async createCovenant(
    counterparty: string,
    terms: string,
    paymentSchedule: PaymentSchedule
  ): Promise<string> {
    if (!this.smartAccount) {
      throw new Error('Smart account not initialized');
    }

    const userOpHash = await this.smartAccount.sendUserOperation({
      calls: [{
        to: this.sdk.addresses.covenantFactory,
        data: this.sdk.covenantFactory.interface.encodeFunctionData('createCovenant', [
          counterparty,
          terms,
          paymentSchedule,
        ]),
      }],
    });

    return userOpHash;
  }

  async setupRecurringPaymentSession(
    covenantId: string,
    covenantAddress: string,
    maxAmount: bigint,
    durationDays: number,
    paymentToken: string
  ): Promise<SessionKey> {
    const validUntil = Math.floor(Date.now() / 1000) + (durationDays * 24 * 60 * 60);

    return {
      sessionKeyAddress: await this.smartAccount.getAddress(),
      sessionData: JSON.stringify({
        covenantId,
        covenantAddress,
        maxAmount: maxAmount.toString(),
        paymentToken,
        validUntil,
      }),
      validUntil,
    };
  }

  async processRecurringPayment(
    covenantAddress: string,
    paymentToken: string,
    amount: bigint
  ): Promise<string> {
    const userOpHash = await this.smartAccount.sendUserOperation({
      calls: [{
        to: covenantAddress,
        data: this.sdk.covenantFactory.interface.encodeFunctionData(
          'processRecurringPayment',
          [paymentToken, amount]
        ),
      }],
    });

    return userOpHash;
  }
}

File: sdk/python/covenant_sdk/accounts/smart_account.py
"""Smart account support for COVENANT backend operations."""

from typing import Optional, Dict, Any
from dataclasses import dataclass


@dataclass
class PaymentSchedule:
    frequency: str  # 'daily' | 'weekly' | 'monthly'
    amount: int
    token: str
    start_time: int
    end_time: int


@dataclass
class SessionKey:
    session_key_address: str
    session_data: str
    valid_until: int


class CovenantSmartAccount:
    """Smart account support for COVENANT backend operations."""
    
    def __init__(self, sdk: 'CovenantSDK', session_key: Optional[str] = None):
        self.sdk = sdk
        self.session_key = session_key
        
    def create_covenant_gasless(
        self,
        counterparty: str,
        terms: str,
        payment_schedule: PaymentSchedule
    ) -> str:
        """Create covenant via user operation (gasless).
        
        Returns userOp hash that can be tracked.
        """
        # Build user operation
        # Submit to bundler via Pimlico API
        # Return userOp hash
        pass
        
    def validate_session_key(self, session_key: str, action: str) -> bool:
        """Validate if session key has permission for action."""
        # Check session key permissions against policy
        # Verify session hasn't expired
        pass
        
    def process_recurring_payment(
        self,
        covenant_address: str,
        payment_token: str,
        amount: int,
        session_key: Optional[str] = None
    ) -> str:
        """Process a recurring payment using session key."""
        # Validate session key permissions
        # Build user operation
        # Submit to bundler
        pass
        
    def get_user_operation_status(self, user_op_hash: str) -> Dict[str, Any]:
        """Check status of a user operation."""
        # Query bundler for receipt
        pass
